ManalHamdi/assignment3/Assignment3.cpp

```diff
@@ -0,0 +1,140 @@
+		for (int i = 0; i < grid.number_rows; i++) {
+			for (int j = 0; j < grid.number_columns; j++) {
+				grid.grid[i][j] = tolower(grid.grid[i][j]);
+			}
+		}
+	}
+	bool validIndex(int row, int column) {
+		return row >= 0 && row < number_rows && column >= 0 && column < number_columns;
+	}
+	std::vector<std::pair<int,int> > getNeighboorIndices(int current_index_row, int current_index_column) {
+		std::vector<std::pair<int,int>> neighboors_indeces;
+		for (int i = -1; i <= 1; i++) {
+			for (int j = -1; j <= 1; j++) {
+				if (validIndex(current_index_row + i, current_index_column + j)) {
+					neighboors_indeces.push_back(std::make_pair(current_index_row + i, current_index_column + j));
+				}
+			}
+		}
+		return neighboors_indeces;
+	}
+};
+
+
+void DFS(Dictionary &dictionary,Grid &grid, std::vector<std::string> &words_found, int current_index_row, int current_index_column, std::string &word, std::vector<std::vector<bool> > &visited) {
+	if(visited[current_index_row][current_index_column]) {
+		return;
+	}
+	visited[current_index_row][current_index_column] = true;
+	word.push_back(grid.grid[current_index_row][current_index_column]);
+	if (dictionary.isPrefix(word)) {
++		if (dictionary.isWord(word)) {
+			words_found.push_back(word);
+		}
+	}
+	else {
+		return;
+	}
+	std::vector<std::pair<int,int>> neighboors_indeces = grid.getNeighboorIndices(current_index_row, current_index_column);
+	for (auto neighboor: neighboors_indeces) {
+		return DFS(dictionary, grid, words_found, neighboor.first, neighboor.second, word, visited);
+	}
+	visited[current_index_row][current_index_column] = false;
+	word.pop_back();
+}
+
+//Assumption: the function should be case insensitive because in real world "eat" and "EAT" stand for the same words
+std::vector<std::string> findAllWordsInGrid(Dictionary &dictionary, Grid &grid) {
+	std::vector<std::string> words_grid;
+	std::string word;
+	std::vector<std::vector<bool> > visited(grid.getNumberRows(), std::vector<bool>(grid.getNumberColumns(), false));
+	
+	grid.validateGrid(grid);
+	
+	for (int i = 0; i < grid.getNumberRows(); i++) {
+		for (int j = 0; j < grid.getNumberColumns(); j++) {
+			DFS(dictionary, grid, words_grid,i,j, word, visited);
+		}
+	}
+	return words_grid;
+}
+class Test {
+public:
+	void testEmptyGrid() {
+		Grid grid((std::vector<std::vector<char>>()));
+		Dictionary dictionary((std::set<std::string>("caa", "notfound", "rdct")));
+		EXPECT_EQ({},findAllWords(dictionary, grid));
+	}
+	void testEmptyDictionary() {
+		Grid grid((std::vector<std::vector<char>>({{'a','a','r'},{'t','c','d'}})));
+		Dictionary dictionary((std::set<std::string>()));
+		EXPECT_EQ({},findAllWords(dictionary, grid));
+	}
+	void testPositive() {
+		Grid grid((std::vector<std::vector<char>>({{'a','a','r'},{'t','c','d'}})));
+		Dictionary dictionary((std::set<std::string>("caa", "notfound", "rdct")));
+		EXPECT_EQ({“caa”, “rdct”},findAllWords(dictionary, grid));
+	}
+};
+int main (void) {
+	Test test = new Test();
```

Comment:
Again, every time you use `new` to create a variable, you *need* to destroy it with `delete test;`, otherwise you're leaking memory. 

Your syntax is very Java-like to me :)
In C++, there's nothing wrong with creating variables like this(and not leaking memory):

`Test test;`

Please look into variable creation using `new` vs locally defined variables in C++, you're making this mistake over and over :)

Reply:
Yes your comment is very right! I was exposed to object oriented only with Java, so it was my reference. I will make the changes and read more about it. Thanks!